File: 127-word-ladder/word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Queue<String> q = new ArrayDeque<>();
        Set<String> seen = new HashSet<>();
        Set<String> set = new HashSet<>(wordList);
        if (!set.contains(endWord)) return 0;
        q.add(beginWord);
        int steps = 1;
        while (!q.isEmpty()) {
            int size = q.size();
            for (int k = 0; k < size; k++) {
                String popped = q.poll();
                seen.add(popped);
                if (popped.equals(endWord)) return steps;
                for (int i = 0; i < popped.length(); i++) {
                    for (int j = 0; j < 26; j++) {
                        String next = popped.substring(0, i) + (char) ('a' + j) + popped.substring(i + 1);
                        if (!seen.contains(next) && set.contains(next)) {
                            seen.add(next);
                            q.add(next);
                        }
                    }
                }
            }
            steps++;
        }
        return 0;
    }

    private boolean isAdjacent(String word1, String word2) {
        int mismatches = 0;
        for (int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) {
                mismatches++;
                if (mismatches > 1) return false;
            }
        }
        return true;
    }
}

File: 127-word-ladder/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        q = deque([beginWord])
        visited = {beginWord}
        wordSet = set(wordList)
        steps = 1

        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                if curr == endWord:
                    return steps
                for j in range(len(beginWord)):
                    for k in range(26):
                        newWord = curr[:j] + chr(ord('a') + k) + curr[j + 1:]
                        if newWord not in visited and newWord in wordSet:
                            q.append(newWord)
                            visited.add(newWord)
            steps += 1
        return 0
